### PR TITLE
Server Push caveats

### LINKED ISSUE #786: Cut Server Push
@ianswett says:

It's not very widely used and many implementations either don't support it at all or don't support it in a way that improves end-user performance.

This sparked a lively discussion at https://github.com/martinthomson/http2v2/issues/16.

### PR DESCRIPTION
**Pull request by mnot**
Fixes #786

### CODE REVIEW

**Review by Lukasa — APPROVED** [review]
LGTM

File: draft-ietf-httpbis-http2bis.xml [hunk 1]
```diff
@@ -3216,9 +3215,20 @@
...
-          client-initiated request. This can be useful when the server knows the client will need
-          to have those responses available in order to fully process the response to the original
-          request.
+          client-initiated request.
+        </t>
+        <t>
+          Server push was designed to allow a server to exploit knowledge about the resources it
+          serves, along with typical client behavior, to improve client-perceived performance, by
+          reducing the latency associated with content that links to other content that needs to be
+          fetched to present it (for example, stylesheets and scripts, in the case of HTML).
```
Comment (martinthomson) [inline draft-ietf-httpbis-http2bis.xml:3224]: This is one mega-sentence.  How about:

```suggestion
          Server push was designed to allow a server to exploit knowledge about resources to
          improve client-perceived performance.  A server that could correctly predict a request
          might reduce latency by pushing a response to clients. For example, with HTML, server
          push might be used for linked content like stylesheets, scripts, and images.
```

**Review by martinthomson — APPROVED** [review]
This is great.  I had some trouble with the mega-sentence.  That said, my rewording might not be much better.

File: draft-ietf-httpbis-http2bis.xml (hunk 1 above)
Comment (mnot) [reply draft-ietf-httpbis-http2bis.xml]: PTAL at latest commit

File: draft-ietf-httpbis-http2bis.xml [hunk 2]
```diff
@@ -3216,9 +3215,23 @@
...
-          client-initiated request. This can be useful when the server knows the client will need
-          to have those responses available in order to fully process the response to the original
-          request.
+          client-initiated request.
+        </t>
+        <t>
+          Server push was designed to allow a server to improve client-perceived performance by
+          predicting what requests will follow those that it receives, thereby removing a round
+          trip for them. For example, a request for HTML is often followed by requests
+          for stylesheets and scripts associated with that page. When these requests
```
Comment (martinthomson) [inline draft-ietf-httpbis-http2bis.xml:3224]:
```suggestion
          for stylesheets and scripts referenced by that page. When these requests
```

File: draft-ietf-httpbis-http2bis.xml [hunk 3]
```diff
@@ -3216,9 +3215,23 @@
...
+          are pushed, the client does not need to wait to receive the references to them in the HTML
+          and issue separate requests.
+        </t>
+        <t>
+          In practice, server push is difficult to use effectively, because it requires the
+          server to correctly anticipate the additional requests the client will make, taking into
+          account factors such as caching, content negotiation, and user behavior. Errors in
+          prediction can lead to performance degradation, due to the opportunity cost that the
+          additional data on the wire represents. In particular, pushing more than a round trip of
+          data can cause contention issues with more-important responses.
```
Comment (martinthomson) [inline draft-ietf-httpbis-http2bis.xml:3234]: I think that you might say "pushing any significant amount of data".  It doesn't need to be a whole round trip worth.  Even a fraction of a round trip worth can starve out a high priority request that arrives immediately after starting.

**Review by martinthomson — APPROVED** [review]
Very nice.  Thanks Mark.

**Pull request merged**